### PR TITLE
cmake: enable C++11 implicitly

### LINKED ISSUE #11428: Tutorial "Using OpenCV with gcc and CMake" doesn't build in 4.0 - set C++11 in CMake
##### System information (version)

- OpenCV => 4.0
- Operating System / Platform => Mac OSX 10.13 64 Bit
- Compiler => Apple LLVM version 9.1.0 (clang-902.0.39.1)

##### Detailed description

The introductory tutorial at this [page](https://docs.opencv.org/master/db/df5/tutorial_linux_gcc_cmake.html) fails to build as presented in the tutorial. Running `$ make` generates `error: "OpenCV 4.x+ requires enabled C++11 support"`. The fix on my system seems simple as demonstrated by this `git diff`:
```
--- a/doc/tutorials/introduction/linux_gcc_cmake/linux_gcc_cmake.markdown
+++ b/doc/tutorials/introduction/linux_gcc_cmake/linux_gcc_cmake.markdown
@@ -51,6 +51,8 @@ int main(int argc, char** argv )
 Now you have to create your CMakeLists.txt file. It should look like this:
 @code{.cmake}
 cmake_minimum_required(VERSION 2.8)
+set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
 project( DisplayImage )
 find_package( OpenCV REQUIRED )
 include_directories( ${OpenCV_INCLUDE_DIRS} )
``` 

If this is a time-worthy fix I'd be happy to create a PR with the change above.

##### Steps to reproduce

Follow the tutorial, creating the `DisplayImage.cpp` and `CMakeLists.txt` files. Run `cmake .` and `make` also as specified in the tutorial.

To fix the issue, add the two lines specified in `git diff` above to the `CMakeLists.txt` file.

### LINKED ISSUE #13000: Undefined reference to cv::imread(std::string const&, int)
I am building a project with OpenCV 4 beta and when I use the function cv::imread or cv::imwrite I get an undefined reference to the functions.

My CMakeLists.txt is as follows:

cmake_minimum_required(VERSION 3.1.0)

project(myImageProgram)

set(CMAKE_PREFIX_PATH /home/marios-cellink/libtorch)

find_package(OpenCV 4.0.0 REQUIRED)
find_package(Torch REQUIRED)
add_executable(${PROJECT_NAME} "main.cpp")

message(STATUS "OpenCV library status:")
message(STATUS "    config: ${OPENCV_DIR}")
message(STATUS "    version: ${OpenV_VERSION}")
message(STATUS "    libraries: ${OpenCV_LIBS}")
message(STATUS "    include_path: ${OpenCV_INCLUDE_DIRS}")

message(STATUS "Torch library status:")
message(STATUS "    include_path: ${TORCH_INCLUDE_DIRS}")
message(STATUS "    TORCHLIB: ${TORCH_LIBRARIES}")

target_link_libraries(${PROJECT_NAME} ${OpenCV_LIBS})
target_link_libraries(${PROJECT_NAME} ${TORCH_LIBRARIES})

And the program is as follows:

#include <iostream>
#include <torch/script.h>
#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/imgcodecs.hpp>

int main()
{
    cv::Mat orig = cv::imread("../myImageProgram/0003_.png");
    cv::Mat img = cv::imread("../myImagePorgram/0003_.png", cv::IMREAD_COLOR);
    std::shared_ptr<torch::jit::script::Module> module = torch::jit::load("../myImageProgram/model.pt");
    return 0;
}

Am I linking OpenCV correctly with cmake? It seems that if I comment out the torch parts from CMake, the program build succesfully. Any ideas?

### PR DESCRIPTION
**Pull request by alalek**
resolves #13000
resolves #11428

```
build_image:Custom=centos:7
buildworker:Custom=linux-1
build_contrib:Custom=OFF
```

### CODE REVIEW

File: cmake/templates/OpenCVConfig.cmake.in [hunk 1]
```diff
@@ -273,6 +273,22 @@ endif()
...
+    message(STATUS "OpenCV: CMake version is low (${CMAKE_VERSION}, required 3.1+). Can't enable C++11 features: https://github.com/opencv/opencv/issues/13000")
+  endif()
+else()
+  set(__target opencv_core)
+  if(TARGET opencv_world)
+    set(__target opencv_world)
+  endif()
+  # Simulate exported result of target_compile_features(opencv_core PUBLIC cxx_auto_type)
+  set_target_properties(${__target} PROPERTIES
+    INTERFACE_COMPILE_FEATURES "cxx_auto_type"
```
Comment (sergiud) [inline cmake/templates/OpenCVConfig.cmake.in]: I wouldn't list only `cxx_auto_type` but also other C++11 features actually used in OpenCV's interface (e.g., `cxx_rvalue_references` etc.) Optionally, one could replace these feature requirements by `cxx_std_11` given CMake version is 3.8 or above.

Comment (alalek) [reply cmake/templates/OpenCVConfig.cmake.in]: Added CMake 3.8+ handling through `cxx_std_11`.

Extended list: `cxx_auto_type cxx_rvalue_references cxx_lambdas`
These features are mostly used in OpenCV public headers.

Comment (sergiud) [reply cmake/templates/OpenCVConfig.cmake.in]: LGTM.

**Comment by alalek** [comment]
:+1:

**Pull request merged**